File: be/src/system/capabilities.ts
import { execFile } from 'child_process';
import * as os from 'os';
import { z } from 'zod';

export const CAPABILITIES = 'CAPABILITIES';

export interface Capabilities {
  platform: string;
  arch: string;
  chip: string | null;
  memory_gb: number;
  gpu_eligible: boolean;
  recommended_preset: 'light' | 'standard' | 'quality' | null;
}

export function buildCapabilities(input: {
  platform: string;
  arch: string;
  totalmemBytes: number;
  chip: string | null;
  /** 실측값이 있을 때만 — 없으면 platform/arch에서 추정한다. */
  gpuEligible?: boolean;
}): Capabilities {
  const gpuEligible = input.gpuEligible ?? (input.platform === 'darwin' && input.arch === 'arm64');
  const memoryGb = Math.round(input.totalmemBytes / 1024 ** 3);
  // 모든 프리셋이 diarization gpu를 포함 — 비적격 환경엔 추천 불가 (spec §3)
  const recommended = !gpuEligible
    ? null
    : memoryGb < 16
      ? ('light' as const)
      : memoryGb < 48
        ? ('standard' as const)
        : ('quality' as const);
  return {
    platform: input.platform,
    arch: input.arch,
    chip: input.chip,
    memory_gb: memoryGb,
    gpu_eligible: gpuEligible,
    recommended_preset: recommended,
  };
}

function sysctlChip(): Promise<string | null> {
  return new Promise((resolve) => {
    execFile('sysctl', ['-n', 'machdep.cpu.brand_string'], { timeout: 1000 }, (err, stdout) => {
      resolve(err ? null : stdout.trim() || null);
    });
  });
}

let cached: Capabilities | null = null;
export async function detectCapabilities(): Promise<Capabilities> {
  if (cached) return cached;
  // 스펙 전체가 워커가 도는 머신의 것이지 API가 도는 머신의 것이 아니다. 배포 이미지
  // (deploy/)에서는 API가 Linux 컨테이너 안이고 워커는 호스트 Mac이라, compose가
  // 호스트 값을 이 env들로 알려준다. 개발에선 비어 있고 프로세스 값을 쓴다.
  const platform = process.env.CAPABILITIES_PLATFORM || process.platform;
  const arch = process.env.CAPABILITIES_ARCH || process.arch;
  // 컨테이너 안의 os.totalmem()은 Docker Desktop VM 할당량이지 Mac의 RAM이 아니다 —
  // 그대로 두면 "내 머신" 카드가 작은 값을 보이고 추천 프리셋이 light로 눌러앉는다.
  const envMemoryGb = Number(process.env.CAPABILITIES_MEMORY_GB);
  const totalmemBytes =
    Number.isFinite(envMemoryGb) && envMemoryGb > 0 ? envMemoryGb * 1024 ** 3 : os.totalmem();
  // sysctl 실행 가능 여부는 실제 프로세스 플랫폼이 정한다 — env의 darwin은 호스트
  // 주장일 뿐 이 프로세스가 macOS라는 뜻이 아니다 (컨테이너엔 sysctl이 없다).
  const chip =
    process.env.CAPABILITIES_CHIP || (process.platform === 'darwin' ? await sysctlChip() : null);
  cached = buildCapabilities({
    platform,
    arch,
    totalmemBytes,
    chip,
  });
  return cached;
}

export const WORKER_CAPABILITIES_KEY = 'worker_capabilities';

/**
 * 워커가 `app_setting`에 적어두는 자기 머신 관측값 — Python 쪽
 * `damwha_worker/capabilities.py`와 짝을 이루는 계약이다. job 페이로드와 달리
 * **단방향**(워커 write / API read)이라 pydantic 짝이 없고 zod만 있다. 워커가 나중에
 * 필드를 더 붙여도 API가 깨지지 않도록 strict가 아니다 (진단용 `gpu_probe`가 그렇게
 * 실려 오고, 여기서 조용히 버려진다).
 */
export const WorkerCapabilitiesSchema = z.object({
  worker_id: z.string(),
  platform: z.string(),
  arch: z.string(),
  chip: z.string().nullable(),
  memory_gb: z.number().int().positive(),
  gpu_eligible: z.boolean(),
});
export type WorkerCapabilities = z.infer<typeof WorkerCapabilitiesSchema>;

/**
 * 워커 보고가 있으면 그것이 진실이다 — GPU를 실제로 쓰는 머신이 워커가 도는 머신이고,
 * `gpu_eligible`도 거기서 MPS를 재본 결과다. env 추정과 달리 Rosetta python을 걸러낸다.
 */
export function fromWorkerReport(r: WorkerCapabilities): Capabilities {
  return buildCapabilities({
    platform: r.platform,
    arch: r.arch,
    chip: r.chip,
    totalmemBytes: r.memory_gb * 1024 ** 3,
    gpuEligible: r.gpu_eligible,
  });
}

File: be/worker/damwha_worker/capabilities.py
"""이 워커가 도는 머신의 실제 스펙 — API는 이걸 직접 볼 수 없다.

API의 `detectCapabilities()`는 platform/arch를 env(`CAPABILITIES_*`)로 받아 거기서
`gpu_eligible`을 **추측**한다. 배포 이미지에서 API는 Linux 컨테이너 안이고 워커는
호스트 Mac이라 그 방법 말고는 없지만, 추측은 Rosetta python으로 깔린 워커를 걸러내지
못한다 — `CAPABILITIES_ARCH=arm64`를 그대로 통과해 UI가 GPU 프리셋을 열어주고,
업로드가 처리 도중 `gpu_unavailable`(PERMANENT)로 죽는다.

여기서 보고하는 값은 추측이 아니다. `platform.machine()`은 Rosetta 아래에서 x86_64를
돌려주고, MPS 가용성은 `torch_device`가 쓰는 바로 그 술어(`models.device.mps_available`)를
**자식 프로세스에서 실제로 호출해** 확인한다. 자식으로 미루는 이유는 부모 supervisor가
가벼워야 하기 때문이다 — torch를 부모가 import하면 프로세스 수명 내내 수백 MB를 쥔다.

보고는 `app_setting`의 `worker_capabilities` 키에 들어간다. job 테이블 계약과 달리
**단방향**(워커 write / API read)이라 TypeScript 짝은 zod 스키마 하나뿐이다.
"""

from __future__ import annotations

import os
import platform
import subprocess
import sys

# Node의 process.arch 어휘로 맞춘다 — 같은 필드를 API가 자기 추정으로도 채우므로
# 두 출처가 다른 문자열을 쓰면 화면에 뜨는 값이 출처마다 달라진다.
_ARCH_ALIASES = {"x86_64": "x64", "AMD64": "x64", "aarch64": "arm64"}

_SYSCTL_TIMEOUT_SECONDS = 5.0
# 대부분 torch import 시간이다. 콜드 캐시에서도 넉넉하게.
_PROBE_TIMEOUT_SECONDS = 120.0

_PROBE_CODE = "from damwha_worker.models.device import mps_available; print(int(mps_available()))"


def _sysctl(name: str) -> str | None:
    try:
        r = subprocess.run(
            ["sysctl", "-n", name],
            capture_output=True,
            text=True,
            timeout=_SYSCTL_TIMEOUT_SECONDS,
        )
    except (OSError, subprocess.TimeoutExpired):
        return None
    return r.stdout.strip() or None if r.returncode == 0 else None


def _memory_bytes() -> int | None:
    if sys.platform == "darwin":
        raw = _sysctl("hw.memsize")
        return int(raw) if raw and raw.isdigit() else None
    try:
        return os.sysconf("SC_PAGE_SIZE") * os.sysconf("SC_PHYS_PAGES")
    except (OSError, ValueError):
        return None


def probe_mps() -> bool | None:
    """`torch_device`와 같은 술어를 자식 프로세스에서 호출한다. 판정 불가면 None.

    None은 torch 미설치(모델 extra 없음)나 프로브 자체 실패를 뜻한다. 그 경우 GPU
    자격을 뺏지 않고 아키텍처 판단으로 돌아간다 — 잡을 아예 못 도는 워커가 GPU
    프리셋까지 잠글 이유는 없고, 그 고장은 다른 곳에서 드러난다.
    """
    try:
        r = subprocess.run(
            [sys.executable, "-c", _PROBE_CODE],
            capture_output=True,
            text=True,
            timeout=_PROBE_TIMEOUT_SECONDS,
        )
    except (OSError, subprocess.TimeoutExpired):
        return None
    if r.returncode != 0:
        return None
    return {"1": True, "0": False}.get(r.stdout.strip())


def detect(worker_id: str) -> dict:
    """app_setting에 그대로 실릴 dict. 물리 메모리를 못 읽으면 보고를 포기한다.

    부분 보고(메모리 0)를 올리면 API가 그걸 진실로 받아 추천 프리셋을 엉뚱하게 잡는다 —
    아무것도 안 올려서 API가 자기 추정으로 폴백하는 편이 낫다.
    """
    memory_bytes = _memory_bytes()
    if not memory_bytes:
        raise RuntimeError("could not read physical memory size on this machine")
    machine = _ARCH_ALIASES.get(platform.machine(), platform.machine())
    # Rosetta 아래에서는 machine()이 x86_64다 — env 추측과 갈리는 지점이 이 한 줄이다.
    apple_silicon = sys.platform == "darwin" and machine == "arm64"
    mps = probe_mps()
    return {
        "worker_id": worker_id,
        "platform": sys.platform,
        "arch": machine,
        "chip": _sysctl("machdep.cpu.brand_string") if sys.platform == "darwin" else None,
        "memory_gb": round(memory_bytes / 1024**3),
        "gpu_eligible": apple_silicon if mps is None else (apple_silicon and mps),
        # 잠긴 GPU 프리셋을 psql 한 줄로 설명하기 위한 진단 흔적. API 응답엔 안 나간다.
        "gpu_probe": {True: "mps_available", False: "mps_unavailable", None: "unknown"}[mps],
    }
